src/collectors/block_collector.rs

```diff
@@ -32,37 +33,83 @@ impl<M> BlockCollector<M> {
+
+        let provider = self.provider.clone();
+        let polling_interval = BLOCK_POLLING_INTERVAL;
+
         let stream = async_stream::stream! {
+            let mut last_block = start_block;
+
             loop {
-                if let Some(block_hash) = watcher.next().await {
-                    match self.provider.get_block(block_hash).await {
-                        Ok(Some(block)) => {
-                            yield NewBlock {
-                                hash: block.hash.unwrap(),
-                                number: block.number.unwrap(),
-                                timestamp: block.timestamp,
-                            };
+                // Attempt to watch new blocks
+                let mut watcher = match provider.watch_blocks().await {
+                    Ok(w) => {
+                        info!("Successfully created new block watcher.");
+                        w.interval(polling_interval).stream()
+                    },
+                    Err(e) => {
+                        error!("Failed to create block watcher: {}. Retrying in 5 seconds...", e);
+                        tokio::time::sleep(Duration::from_millis(100)).await;
+                        continue;
+                    }
+                };
+
+                // Iterate over incoming block hashes
+                loop {
+                    match watcher.next().await {
+                        Some(block_hash) => {
+                            match provider.get_block(block_hash).await {
+                                Ok(Some(block)) => {
+                                    let block_number = block.number.unwrap().as_u64();
+                                    let block_timestamp = block.timestamp;
+                                    
+                                    // Update last processed block number
+                                    if block_number > last_block {
+                                        last_block = block_number;
+                                        
+                                        yield NewBlock {
+                                            hash: block.hash.unwrap(),
+                                            number: U64::from(block_number),
+                                            timestamp: block_timestamp,
+                                        };
+                                    }
+                                },
+                                Ok(None) => {
+                                    warn!("Received block hash {} but block not found.", block_hash);
+                                },
+                                Err(e) => {
+                                    error!("Error fetching block {}: {}.", block_hash, e);
+                                }
+                            }
+                        },
+                        None => {
+                            warn!("Block watcher stream ended unexpectedly. Recreating watcher...");
```

Comment:
possible to emit CW metric here?

Reply:
yeah but prefer to do as follow-up item! Also I think the new block metric kinda already serves the same purpose